src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArray = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+
+        break;
+
+      default:
+        break;
+    }
+    stateArray.push({ ...newState });
```

Comment:
You are correctly creating a new object for each state change to avoid mutating the original state. However, it would be more efficient to push the new state object into the `stateArray` inside each case block right after the state modification is done. This way, you avoid repeating the same action after each case and ensure that the state is only copied once per action.